packages/lit-dev-content/site/articles/article/lit-with-tailwind.md

```diff
@@ -0,0 +1,228 @@
+---
+title: Lit components with tailwind styles
+publishDate: 2023-12-10
+lastUpdated: 2023-12-10
+summary: Using tailwind styles inside Lit components
+tags:
+  - web-components
+  - tailwind
+  - css
+eleventyNavigation:
+  parent: Articles
+  key: Lit with tailwind
+  order: 0
+author:
+  - james-garbutt
+---
+
+Tailwind and many other CSS libraries were not designed with web components in
+mind, and come with some non-obvious difficulties when trying to use them in
+such a codebase.
+
+This is a brief guide on how to use such a library with your Lit components.
+
+## Why tailwind doesn't work out of the box
+
+Out of the box, tailwind basically provides some global CSS classes and injects
+the associated CSS at build time (via postcss).
+
+This conflicts with how web components work, since each web component has its
+own natively scoped stylesheet rather than inherting any global styles. This
+is why Tailwind will not be much use to us without further setup.
+
+## Overview of solution
+
+To solve the gaps tailwind comes with, we will:
+
+- Inject tailwind styles into regular CSS files
+- Use [import attributes](https://github.com/tc39/proposal-import-attributes)
+to import those CSS files
+- Use [esbuild](https://github.com/evanw/esbuild) to pull those CSS files
+into the bundle or same output directory

```

Comment:
So I'm not too familiar with features in this area and this is more for my understanding of what's going on.

Looking at the output from the example repo, esbuild produces this line in the bundle
```js
import styles from "./hello-world-4K4DCJGM.css" assert { type: "css" };
```

esbuild is being used to bundle and move the css files into the outdir and is it then left up to the users to make this work on unsupported browsers? We probably need a loud caveat stating this only works in Chrome if that's the case. Can `es-module-shims` be used to make it work for other browser?

Might there be rollup plugins that work with the new correct import attribute syntax?

Reply:
hmmm... latest esbuild actually does support import attributes with the `with` keyword. https://github.com/evanw/esbuild/releases/tag/v0.19.7

https://github.com/evanw/esbuild/issues/3384#issuecomment-1823412482 Evan does say `type: 'css'` support will come to esbuild eventually but perhaps a plugin can be written for it in the mean time?

Reply:
yeah pretty much, right now it just outputs the source as-is (i.e. your browser needs to support import assertions at least, ideally attributes too).

the esbuild blockage is because of no support for CSS. you're right though, we may be able to use a plugin, i'll have a dig around

it'd be nice to not have to use rollup, just because it consumes most of the build whereas esbuild focuses on only the bundling (as in we should also do the tailwind preprocessing in rollup, if we use rollup).

Reply:
that makes sense. i don't disagree with the tool choice and i don't want to block or dictate too much as to what this article should say.

since this article being the first of its kind, i think it would be good to convey to the reader that this is meant as _one_ way to solve this problem, with caveats like browser support, and mention alternatives (even without details) like using rollup plugins for all the steps. being published on lit.dev, people might consider this the only recommended way otherwise.

Reply:
Heya, seeing how `with` is shipped in chrome now + TS 5.4 supporting it, can we switch to `with`?